test/jdk/jdk/jfr/event/oldobject/TestEmergencyDumpAtOOM.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2025, NTT DATA.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.jfr.event.oldobject;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import jdk.jfr.consumer.RecordingFile;
+
+import jdk.test.lib.Asserts;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+/**
+* @test
+* @bug 8364090
+* @summary Tests Dump reason and OldObjectSample events at OOME.
+* @requires vm.flagless
+* @requires vm.hasJFR
+* @library /test/lib
+* @run main/othervm jdk.jfr.event.oldobject.TestEmergencyDumpAtOOM
+*/
+public class TestEmergencyDumpAtOOM {
+
+    public static List<String> DEFAULT_LEAKER_ARGS = List.of(
+        "-Xmx32m",
+        "-XX:StartFlightRecording:path-to-gc-roots=true,dumponexit=true,filename=oom.jfr",
+        Leaker.class.getName()
+    );
+
+    public static class Leaker {
+        public static void main(String... args) {
+            List<byte[]> list = new ArrayList<>();
+            while (true) {
+                list.add(new byte[1024]);
+            }
+        }
+    }
+
+    private static void test(boolean shouldCrash) throws Exception {
+        var args = new ArrayList<String>(DEFAULT_LEAKER_ARGS);
+        if (shouldCrash) {
+            args.add(0, "-XX:+CrashOnOutOfMemoryError");
+        }
+        var p = ProcessTools.createTestJavaProcessBuilder(args).start();
+        p.waitFor();
+        var output = new OutputAnalyzer(p);
```

Comment:
I'm not a big fan of var, since it degrades the readability. I like to know which types I deal with when reading code.

Not sure what others think, though, or if there is a Java style guideline similar to our hotspot style guide.

Reply:
Using an EventStream might simplify the filtering, e.g.

        AtomicLong oldObjects = new AtomicLong();
        AtomicReference<String> reason = new AtomicReference<>();
        try (EventStream stream = EventStream.openFile(...)) {
            stream.onEvent("jdk.Shutdown", e -> reason.set(e.getString("reason")));
            stream.onEvent("jdk.OldObjectSample", e -> oldObjects.incrementAndGet());
            stream.start();
        }
        if (shouldCrash) {
            // validation
        }

Reply:
I updated the test. I hope it works on your environment.

- Revise JVM option for the test process (`-Xmx` and `-XX:TLABSize` are copied from TestClassLoaderLeak.java)
- Use `EventStream` to check JFR events as Erik shown
- Repeat the test if `OldObjectSample` does not appear on the flight record
- Remove `var` from test code

Reply:
I will try running it. I think readability can be improved if everything under shouldCrash is grouped together, e.g.

```
if (oldObjects.get() > 0L) {
    if (shouldCrash) {
        Asserts.assertEquals("VM Error", shutdownReason.get());
        Asserts.assertEquals("Out of Memory", dumpReason.get());
    } else {
        Asserts.assertEquals("No remaining non-daemon Java threads", shutdownReason.get());
    }
    return;
}
```

Reply:
@egahlin 
I updated testcase as you shown.